File: artifacts/api-server/src/domain/grade_book_factory.py
"""GradeBookFactory — selects the correct GradeBook subclass at runtime."""
from .grade_book import GradeBook
from .weighted_grade_book import WeightedGradeBook
from .curved_grade_book import CurvedGradeBook
from .pass_fail_grade_book import PassFailGradeBook


class GradeBookFactory:
    @staticmethod
    def create(grading_scheme: str) -> GradeBook:
        if grading_scheme == "curved":
            return CurvedGradeBook()
        if grading_scheme == "pass_fail":
            return PassFailGradeBook()
        return WeightedGradeBook()  # default

File: server/src/domain/GradeBookFactory.ts
import type { GradingScheme } from "@workspace/db";
import { GradeBook } from "./GradeBook";
import { WeightedGradeBook } from "./WeightedGradeBook";
import { CurvedGradeBook } from "./CurvedGradeBook";
import { PassFailGradeBook } from "./PassFailGradeBook";

export class GradeBookFactory {
  static create(scheme: GradingScheme | null | undefined): GradeBook {
    switch (scheme) {
      case "curved":
        return new CurvedGradeBook(5);
      case "pass_fail":
        return new PassFailGradeBook(60);
      case "weighted":
      default:
        return new WeightedGradeBook();
    }
  }
}
